### ISSUE TITLE
UTF-8 BOM characters not supported

### ISSUE THREAD

**Issue by dmedine** [bids-standard/legacy-validator#878]
When a file is formatted in UTF-8 with BOM (which is a standard utf-8 file format), this BIDS validator will issue errors when reading the byte order marker (ï»¿). For example, this is the output:

Location:
BIDS/sub-01/eeg/sub-01_task-DKVisualOddballactiCHampslim4142_channels.tsv

Reason:
Tabular file contains custom columns not described in a data dictionary

Evidence:
Columns: ï»¿name not defined, please define in: /task-DKVisualOddballactiCHampslim4142_channels.json, /channels.json,/sub-01/sub-01_task-DKVisualOddballactiCHampslim4142_channels.json,/sub-01/sub-01_channels.json,/sub-01/eeg/sub-01_task-DKVisualOddballactiCHampslim4142_channels.json,/sub-01/eeg/sub-01_channels.json

**Comment by effigies** [comment]
Sure, we should probably fix this. My extremely limited Javascript knowledge suggests that perhaps we can resolve it by making the following change:

```Diff
     const customColumns = getCustomColumns(
       tsv.contents
+        .replace(/^\uefff/, '')
         .split('\n')[0]
         .trim()
         .split('\t'),
       getTsvType(tsv.file),
     )
```

Here:

https://github.com/bids-standard/bids-validator/blob/f2dd08332748bd476131bd689ea48a85307ec257/bids-validator/validators/tsv/validateTsvColumns.js#L55-L61

I'll note that it's not required or recommended in UTF-8 for the specific reasons that it has no effect on interpretation and it can cause issues with some applications. You may have issues with other BIDS apps besides the validator.

**Comment by dmedine** [comment]
I think that would be for the best. The other validators (e.g. the python based validator) handles the BOM automatically. I am surprised that this needs to be explicitly handled in the javascript function, but so it is. 

It would be great if you could merge this change. Let me know if you need help testing.

**Comment by dmedine** [comment]
I will complain about other BOM related matters if they ever come up ;-)

**Comment by effigies** [comment]
Proposed the change. Should probably note that the Python validator does not check contents, just file names, so I don't know for a certainty that Python ignores these.

**Comment by dmedine** [comment]
Thanks! By our testing this web based validator is the only one complaining about the BOM. But I will report anything if it comes up.

**Comment by effigies** [comment]
Reopening, because this isn't resolved. I can replicate with the dataset shared in https://github.com/bids-standard/bids-validator/pull/879#issuecomment-566597116: https://drive.google.com/open?id=1K4hu7wpV4jmarwCgtgUR_7e5sjTuSqdf

Presumably Javascript has some way to load file contents, so hopefully somebody can figure this out. Loading with Python:

```Python
>>> eventstsv = Path.home() / 'Downloads/BIDSv1/sub-01/eeg/sub-01_task-DKVisualOddballactiCHampslim5142_events.tsv'
>>> eventstsv.read_bytes()[:5]                                                                                                                                                                                                                                                    
b'\xef\xbb\xbfon'
>>> eventstsv.read_text()[:5]                                                                                                                                                                                                                                                    
'\ufeffonse'
>>> eventstsv.read_bytes()[:5].decode('latin-1')                                                                                                                                                                                                                                 
'ï»¿on'
```

So this appears to be an encoding issue where TSV files are read as latin-1, not unicode. That's about as far as I can get.

**Comment by effigies** [comment]
cc @david-nishi

**Comment by effigies** [comment]
Closed in #948. Please reopen if this is still an issue.
